### ISSUE TITLE
Graph break in submodule causes additional graph breaks at submodule boundaries

### ISSUE THREAD

**Issue by fynnsu** [pytorch/pytorch#120009]
### 🐛 Describe the bug

When you have a graph break in a submodule it seems to also produce graph breaks at the start and end of the submodule.

```python
# Submodule called by outer module (outer module is `torch.compile`d)
def forward(self, x):
    # the graph break below results in an additional graph break here
    ...
    torch._dynamo.graph_break()
    ...
    # and here
    return x
```
when one would expect it to produce only a single graph break.

See repro.

### Minified repro

```python
import torch


class HasGraphBreak(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = torch.nn.Linear(5, 5)
        self.linear2 = torch.nn.Linear(5, 5)

    def forward(self, x):
        x = self.linear1(x)
        torch._dynamo.graph_break()
        return self.linear2(x)


class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.submodule = HasGraphBreak()

    def forward(self, x):
        x = torch.relu(x)
        x = self.submodule(x)
        return torch.relu(x)


def custom_backend(gm, _ei):
    _ = gm.print_readable()
    return gm.forward


m = torch.compile(Mod(), backend=custom_backend)

x = torch.randn(3, 5)

_ = m(x)
```
Produces 4 subgraphs (instead of 2):
```python
class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: graph_breaks.py:27, code: x = torch.relu(x)
        x = torch.relu(l_x_);  l_x_ = None
        return (x,)
        
class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: graph_breaks.py:11, code: x = self.linear1(x)
        x = self.L__self___linear1(l_x_);  l_x_ = None
        return (x,)
        
class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: graph_breaks.py:13, code: return self.linear2(x)
        l__self___linear2 = self.L__self___linear2(l_x_);  l_x_ = None
        return (l__self___linear2,)
        
class GraphModule(torch.nn.Module):
    def forward(self, L_stack0_ : torch.Tensor):
        x = L_stack0_
        
        # File: graph_breaks.py:29, code: return torch.relu(x)
        relu = torch.relu(x);  x = None
        return (relu,)
```

### Versions

PyTorch version: 2.2.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.0 (default, Nov 15 2020, 14:28:56)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-6.2.0-37-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 535.113.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 7 3800X 8-Core Processor
CPU family:                         23
Model:                              113
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        4558.8862
CPU min MHz:                        2200.0000
BogoMIPS:                           7800.12
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          256 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           4 MiB (8 instances)
L3 cache:                           32 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] torch==2.2.0
[pip3] torchaudio==2.2.0
[pip3] torchvision==0.17.0
[pip3] torchviz==0.0.2
[pip3] triton==2.2.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] torch                     2.2.0                    pypi_0    pypi
[conda] torchaudio                2.2.0                    pypi_0    pypi
[conda] torchvision               0.17.0                   pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

**Comment by fynnsu** [comment]
Note: this isn't limited to submodules. Also applies to nested fns (and likely nested scopes in general).

**Comment by masnesral** [comment]
Not quite sure who to tag to move this along through triage, so making a wild guess: @anijain2305, is this something we should look into?

**Comment by fynnsu** [comment]
I also wasn't sure if this counts as a bug or if it's the intended behaviour. However, it seems unexpected from a user perspective, and I couldn't think of a good justification for it.

There's potentially an argument to be made for the graph break at the end of the submodule, because if you have a graph break from a branch condition (ex. data dependent control flow), then you will likely have multiple different recompilations of the remainder of the entire graph. Adding a break at the end of the submodule, means you only have to recompile the remainder of the submodule.

Ex. If you have a Module A with a submodule B that is called during A, the execution looks like:
A_before_submodule_call -> B_before_break -> GRAPH BREAK -> B_after_break -> A_after_submodule_call

Then for each branch of the control flow logic, you would need to recompile both "B_after_break" and "A_after_submodule_call" together.

However, with the added graph break:
A_before_submodule_call -> B_before_break -> GRAPH BREAK -> B_after_break -> GRAPH BREAK -> A_after_submodule_call

Only "B_after_break" will be recompiled multiple times and "A_after_submodule_call" will be compiled only once.

However, this only applies to cases where the graph break results in multiple different continuations. Furthermore, even in this case the graph break at the start of the submodule doesn't make sense. And it is unclear to me if the slight overhead reduction obtained by recompiling a smaller graph is worth the loss of fusion/optimizations opportunities another graph break entails.

**Comment by ezyang** [comment]
Dupe of https://github.com/pytorch/pytorch/issues/111003

**Comment by masnesral** [comment]
> Dupe of https://github.com/pytorch/pytorch/issues/111003

@ezyang Does that mean we just close this one?

**Comment by Chillee** [comment]
Yes, closing this in favor of #111003. This has a partial solution (https://github.com/pytorch/pytorch/pull/117906), but I'm not sure if we've committed to solving this.

cc: @jansel @anijain2305
